alpha/veneer/composite/builder.go

```diff
@@ -0,0 +1,317 @@
+		return fmt.Errorf("schema %q does not match the raw veneer builder schema %q", vd.Schema, RawVeneerBuilderSchema)
+	}
+	// Parse out the raw veneer configuration
+	rawConfig := &RawVeneerConfig{}
+	err := json.Unmarshal(vd.Config, rawConfig)
+	if err != nil {
+		return fmt.Errorf("unmarshalling raw veneer config: %w", err)
+	}
+
+	// validate the raw config fields
+	valid := true
+	validationErrs := []string{}
+	if rawConfig.Input == "" {
+		valid = false
+		validationErrs = append(validationErrs, "raw veneer config must have a non-empty input (veneerDefinition.config.input)")
+	}
+
+	if rawConfig.Output == "" {
+		valid = false
+		validationErrs = append(validationErrs, "raw veneer config must have a non-empty output (veneerDefinition.config.output)")
+	}
+
+	if !valid {
+		return fmt.Errorf("raw veneer configuration is invalid: %s", strings.Join(validationErrs, ","))
+	}
+
+	// build the container command
+	containerCmd := exec.Command(rb.builderCfg.ContainerCfg.ContainerTool,
+		"run",
+		"--rm",
+		"-v",
+		fmt.Sprintf("%s:%s", rb.builderCfg.CurrentDirectory, rb.builderCfg.ContainerCfg.WorkingDir),
+		"--entrypoint=cat", // This assumes that the `cat` command is available in the container -- Should we also build a `... render-veneer raw` command to ensure consistent operation? Does OPM already have a way to render a raw FBC?
+		rb.builderCfg.ContainerCfg.BaseImage,
+		path.Join(rb.builderCfg.ContainerCfg.WorkingDir, rawConfig.Input))
+
+	return build(containerCmd, path.Join(dir, rawConfig.Output), rb.builderCfg.OutputType)
+}
+
+func (rb *RawBuilder) Validate(dir string) error {
+	return validate(rb.builderCfg.ContainerCfg, path.Join(rb.builderCfg.CurrentDirectory, dir))
+}
+
+type CustomBuilder struct {
+	builderCfg BuilderConfig
+}
+
+var _ Builder = &CustomBuilder{}
+
+func NewCustomBuilder(builderCfg BuilderConfig) *CustomBuilder {
+	return &CustomBuilder{
+		builderCfg: builderCfg,
+	}
+}
+
+func (cb *CustomBuilder) Build(dir string, vd VeneerDefinition) error {
+	if vd.Schema != CustomVeneerBuilderSchema {
+		return fmt.Errorf("schema %q does not match the custom veneer builder schema %q", vd.Schema, CustomVeneerBuilderSchema)
+	}
+	// Parse out the raw veneer configuration
+	customConfig := &CustomVeneerConfig{}
+	err := json.Unmarshal(vd.Config, customConfig)
+	if err != nil {
+		return fmt.Errorf("unmarshalling custom veneer config: %w", err)
+	}
+
+	// validate the custom config fields
+	if customConfig.Command == "" {
+		return errors.New("custom veneer configuration is invalid: custom veneer config must have a non-empty command (veneerDefinition.config.command)")
+	}
+
+	// build the command to execute
+	// TODO: should the command be run within the container?
+	cmd := exec.Command(customConfig.Command, customConfig.Args...)
+	cmd.Dir = cb.builderCfg.CurrentDirectory
+
+	// TODO: Should we capture the output here for any reason?
+	// Should the custom veneer output an FBC to STDOUT like the other veneer outputs?
```

Comment:
IMHO, we should assume that the custom approach to produce valid FBC in the target destination.  I wouldn't even attempt to capture any files or move anything around.  If folks want to use the custom approach, they really need to be responsible for everything we'd otherwise do.

Reply:
Does the custom script even know what the configured destination is?  Seems like we'd have to feed the destination is an an envvar or something.

Another option is that we say "if exit status is 0, stdout is streamed into a single catalog file in the configured destination". I sort of like this idea because it means that it follows the pattern of other veneers (e.g. output stuff to stdout and let the caller decide what to do with it).

Reply:
> I sort of like this idea because it means that it follows the pattern of other veneers (e.g. output stuff to stdout and let the caller decide what to do with it).

@joelanford This is the reason why I wrote that comment, so that all the builders process everything the exact same way :).

> Does the custom script even know what the configured destination is? Seems like we'd have to feed the destination is an an envvar or something.

Nope, in my demo composite repo I show an example that pipes output to a particular file: https://github.com/everettraven/fbc-composite-example/blob/107df84f1618d95905f32d6708062409c5e2cb3d/catalog/config.yaml#L33-L44

That being said, we could update the custom config spec to require defining an output location and then we process it just like all the other builders.

Reply:
Updated the custom veneer builder to work the same as all the other builders in commit https://github.com/operator-framework/operator-registry/pull/1054/commits/a9e6f68adc310362ea221b29fe56ae7047c98107